### PR TITLE
WIP: kubeadm: don't manually create the Node object during init

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

xref https://github.com/kubernetes/kubernetes/pull/83339

> don't manually create the Node object during init

This proved to be problematic if the kubelet is running within a
cloud-provider. The problem is that the kubeadm created Node object
does not trigger an update in the providerID field under the Node spec.
The reason for that is unclear, yet it hinted that this manual Node
creation is not ideal.

To solve the problem without touching the phase orders:
- Remove the manual node creation (EnsureNodeObject()) and it's calls
in app/phases/* code
- In the cmd/phases/iinit markcontrolplane and uploadconfig phases
before patching the Node object make sure that the "system:nodes" group
is granted the
"system:certificates.k8s.io:certificatesigningrequests:selfnodeclient"
ClusterRole which will allow the kubelet to renew it's own client
certificate and create the Node object.
The same CRB is later created by the boostraptoken init phase, but
the operation will not error out and act as transparent update.

Ideally:
- The bootstraptoken phase should have been before all phases the
patch Nodes.
- The uploadconfig should have not include patching node actions
and this should have been a separate phase.

**Which issue(s) this PR fixes**:

xref  https://github.com/kubernetes/kubernetes/issues/83923
xref https://github.com/kubernetes/kubernetes/pull/83339

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/kind bug
/priority important-soon
/assign @fabriziopandini 
cc @dims

### CODE REVIEW

**Comment by neolit123** [comment]
adding WIP
we need to test if this PR fixes the providerID problem (probably does).

File: cmd/kubeadm/app/util/apiclient/wait.go [hunk 1]
```diff
@@ -135,7 +135,6 @@ func (w *KubeWaiter) WaitForPodToDisappear(podName string) error {
 // WaitForHealthyKubelet blocks until the kubelet /healthz endpoint returns 'ok'
 func (w *KubeWaiter) WaitForHealthyKubelet(initalTimeout time.Duration, healthzEndpoint string) error {
 	time.Sleep(initalTimeout)
-	fmt.Printf("[kubelet-check] Initial timeout of %v passed.\n", initalTimeout)
```
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/wait.go]: this is a separate commit with a description.

Comment (vincepri) [reply cmd/kubeadm/app/util/apiclient/wait.go]: omg thank you, this log line has driven me nuts when debugging GCP 😂

**Comment by dims** [comment]
/retest

@neolit123 this works fine! 

https://pasted.tech/pastes/ecdde300982b465a727867588ac442d68d10165b.raw

Thanks,
Dims

**Comment by ereslibre** [comment]
Thank you @neolit123. Based on @dims's feedback:

/lgtm

> This proved to be problematic if the kubelet is running within a
cloud-provider

It looks like this introduced a time window in which `cluster-api` is watching for the node to appear, and when it does (placeholder preemptly created by kubeadm), it fails because `providerID` is not yet set until the `kubelet` patches the node object.

If we want to keep this approach of  preemptly create this placeholder node object then I guess capi would have to not only wait for the node to exist, but that it also has a `providerID` before calling it done -- unsure if this is feasible, not very familiar with capi nowadays.

**Comment by neolit123** [comment]
thanks @ereslibre 
@dims confirmed that this fixes the problem.

> It looks like this introduced a time window in which cluster-api is watching for the node to appear, and when it does (placeholder preemptly created by kubeadm), it fails because providerID is not yet set until the kubelet patches the node object.

if that is true, it feels to me this might be a bit of oversight on the CAPI side.

> If we want to keep this approach of preemptly creating this placeholder node object then I guess capi would have to not only wait for the node to exist, but that it also has a providerID before calling it done -- unsure if this is feasible, not very familiar with capi nowadays.

yes, it should watch for the node and for the providerID field. so a two condition watch.

the approach in this PR is also not ideal as it requires adding arbitrary RBAC before patching.
i've added this as office hours agenda on Wednesday.

**Comment by neolit123** [comment]
as per the discussion today in the kubeadm office hours - closing this PR and reverting #83339

**Pull request closed without merging**